pkgs/by-name/as/asdf-vm/package.nix

```
@@ -1,89 +1,58 @@
-    #
-    # So we should reshim all installed versions every time shell initialized,
-    # because $out always change
-
-    asdfDir="''${ASDF_DIR:-$HOME/.asdf}"
-    asdfDataDir="''${ASDF_DATA_DIR:-$HOME/.asdf}"
-
-    prevAsdfDirFilePath="$asdfDataDir/.nix-prev-asdf-dir-path"
-
-    if [ -r "$prevAsdfDirFilePath" ]; then
-      prevAsdfDir="$(cat "$prevAsdfDirFilePath")"
-    else
-      prevAsdfDir=""
-    fi
-
-    if [ "$prevAsdfDir" != "$asdfDir" ]; then
-      rm -rf "$asdfDataDir"/shims
-      "$asdfDir"/bin/asdf reshim
-      echo "$asdfDir" > "$prevAsdfDirFilePath"
-    fi
-  '';
-
-  asdfPrepareFile = writeScript "asdf-prepare" ''
-    ASDF_DIR="@asdfDir@"
-
-    source "$ASDF_DIR/asdf.sh"
-    ${asdfReshimFile}
-  '';
-in
-stdenv.mkDerivation rec {
+buildGoModule rec {
   pname = "asdf-vm";
-  version = "0.15.0";
+  version = "0.18.0";
 
   src = fetchFromGitHub {
     owner = "asdf-vm";
     repo = "asdf";
-    rev = "v${version}";
-    sha256 = "sha256-quDgoYi+3hZUEAzXWTHuL5UK1T+4o7+G67w0UzZOjJA=";
+    tag = "v${version}";
+    hash = "sha256-BBd+MiRISjMz2m29nNIakG79Oy1k7bZI/Q24QQNp5CY=";
   };
 
+  vendorHash = "sha256-gzlHXIzDYo4leP+37HgNrz5faIlrCLYA7AVSvZ6Uicc=";
+
   nativeBuildInputs = [
     makeWrapper
     installShellFiles
   ];
 
-  buildInputs = [
-    bash
-    curl
-    git
-  ];
-
-  installPhase = ''
-    mkdir -p $out/share/asdf-vm
-    cp -r . $out/share/asdf-vm
-
-    mkdir -p $out/etc/profile.d
-    substitute ${asdfPrepareFile} $out/etc/profile.d/asdf-prepare.sh \
-      --replace "@asdfDir@" "$out/share/asdf-vm"
-
-    mkdir -p $out/bin
-    makeWrapper $out/share/asdf-vm/bin/asdf $out/bin/asdf \
-      --set ASDF_DIR $out/share/asdf-vm
-
-    installShellCompletion --cmd asdf \
-      --zsh completions/_asdf \
-      --fish completions/asdf.fish \
-      --bash completions/asdf.bash
+  # Tests have additional requirements
+  doCheck = false;
+
+  postInstall = ''
+        # Install profile.d script for environment setup
+        mkdir -p $out/etc/profile.d
+        cat > $out/etc/profile.d/asdf-prepare.sh <<'EOF'
+    export ASDF_DIR="$out"
+    export PATH="''${ASDF_DATA_DIR:-$HOME/.asdf}/shims:$PATH"
+    EOF
+
+        # Wrap the asdf binary to set ASDF_DIR
+        wrapProgram $out/bin/asdf \
+          --set ASDF_DIR $out
+
+        # Generate and install shell completions
+        mkdir -p $out/share/completions
+
+        # Generate completions using the built asdf binary
+        $out/bin/asdf completion bash > $out/share/completions/asdf.bash || true
+        $out/bin/asdf completion zsh > $out/share/completions/_asdf || true
+        $out/bin/asdf completion fish > $out/share/completions/asdf.fish || true
+
+        # Install completions to standard locations
+        installShellCompletion --cmd asdf \
+          --bash $out/share/completions/asdf.bash \
+          --zsh $out/share/completions/_asdf \
+          --fish $out/share/completions/asdf.fish
```

Could've make it a single step:
```suggestion
'' + lib.optionalString (stdenv.buildPlatform.canExecute stdenv.hostPlatform) ''
        installShellCompletion --cmd asdf \
          --bash <($out/bin/asdf completion bash) \
          --fish <($out/bin/asdf completion fish) \
          --zsh <($out/bin/asdf completion zsh)
```
The generation is executed only for native compilation, see https://nixos.org/manual/nixpkgs/stable/#installshellfiles-installshellcompletion-exampleusage-guarded for reference